test/unit/mt_membar.c

```diff
@@ -0,0 +1,189 @@
+#include <stdio.h>
+#include <assert.h>
+#include <pthread.h>
+#include <shmem.h>
+
+/* For systems without the PThread barrier API (e.g. MacOS) */
+#include "pthread_barrier.h"
+
+#define T 2
+#define ITER 100
+
+#ifndef MAX
+#define MAX(A,B)   (((A)>(B)) ? (A) : (B))
+#endif
+
+int shared_dest_1 = 0, shared_dest_2 = 0, result = 0;
+int me, npes, errors = 0, sum_error = 0;
+long lock = 0;
+
+pthread_barrier_t fencebar;
+
+long pSync[_SHMEM_REDUCE_SYNC_SIZE];
+int pWrk[MAX(1, _SHMEM_REDUCE_MIN_WRKDATA_SIZE)];
+
+static void * thread_main(void *arg) {
+    int tid = *(int *) arg;
+    int one = 1, zero = 0;
+    int i;
+
+    /* TEST WAIT */
+    for (i = 0; i < ITER; i++) {
+        if (tid == 0) {
+            shared_dest_1 = 1;
+            shmem_int_wait_until(&shared_dest_1, SHMEM_CMP_EQ, zero);
+            shmem_int_atomic_add(&result, one, me);
+        }
+
+        if (tid == 1) {
+            shmem_int_wait_until(&shared_dest_1, SHMEM_CMP_EQ, one);
+            shmem_int_atomic_add(&result, one, me);
+            shared_dest_1 = 0;
+        }
+    }
+
+    pthread_barrier_wait(&fencebar);
+    if (0 == tid) shmem_barrier_all();
+    pthread_barrier_wait(&fencebar);
+
+    if (tid == 0) {
+        errors += result == T * ITER ? 0 : 1;
+        if (result != T * ITER) {
+            printf("ERROR in WAIT test from %d : result = %d, expected = %d\n", 
+                    me, result, T * ITER);
+        }
+        result = 0;
+	shared_dest_1 = 0;
+    }
+
+    pthread_barrier_wait(&fencebar);
+    if (0 == tid) shmem_barrier_all();
+    pthread_barrier_wait(&fencebar);
+
+    /* TEST FENCE */
+    for (i = 0; i < ITER; i++) {
+        if (tid == 0) {
+            shared_dest_1 += 1;
+            shared_dest_2 += 1;
+            shmem_fence();
+            shmem_int_wait_until(&shared_dest_1, SHMEM_CMP_EQ, zero);
+        }
+
+        if (tid == 1) {
+            shmem_int_wait_until(&shared_dest_2, SHMEM_CMP_EQ, one);
+            shmem_int_atomic_add(&result, shared_dest_1 + shared_dest_2, me);
+            shared_dest_2 = 0;
+            shared_dest_1 = 0;
+            shmem_fence();
+        }
+    }
+
+    pthread_barrier_wait(&fencebar);
+    if (0 == tid) shmem_barrier_all();
+    pthread_barrier_wait(&fencebar);
+
+    if (tid == 0) {
+        errors += result == T * ITER ? 0 : 1;
+        if (result != T * ITER) {
+            printf("ERROR in FENCE test from %d : result = %d, expected = %d\n",
+                    me, result, T * ITER);
+        }
+        result = 0;
+	shared_dest_1 = 0;
+	shared_dest_2 = 0;
```

Comment:
Are there hard tabs in here?  If you use vim, you can get it to highlight hard tabs by adding this to your .vimrc:
```
" Show special representation for hard tabs
set listchars=tab:>-
set list
```